src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,425 @@
+#include "crimson/os/seastore/cached_extent.h"
+#include "crimson/os/seastore/lba_manager.h"
+
+namespace crimson::os::seastore {
+
+class OolExtent : public boost::intrusive_ref_counter<
+  OolExtent, boost::thread_unsafe_counter> {
+public:
+  OolExtent(LogicalCachedExtentRef& lextent)
+    : lextent(lextent) {}
+  void set_ool_paddr(paddr_t addr) {
+    ool_offset = addr;
+  }
+  paddr_t get_ool_paddr() const {
+    return ool_offset;
+  }
+  void persist_paddr() {
+    lextent->set_paddr(ool_offset);
+    ool_offset = P_ADDR_NULL;
+  }
+  bufferptr& get_bptr() {
+    return lextent->get_bptr();
+  }
+  LogicalCachedExtentRef& get_lextent() {
+    return lextent;
+  }
+private:
+  paddr_t ool_offset;
+  LogicalCachedExtentRef lextent;
+};
+
+using OolExtentRef = boost::intrusive_ptr<OolExtent>;
+
+/**
+ * ool_record_t
+ *
+ * Encapsulates logic for building and encoding an ool record destined for
+ * an ool segment.
+ *
+ * Uses a metadata header to enable scanning the ool segment for gc purposes.
+ * Introducing a seperate physical->logical mapping would enable removing the
+ * metadata block overhead.
+ */
+class ool_record_t {
+public:
+  ool_record_t(size_t block_size) : block_size(block_size) {}
+  record_size_t get_encoded_record_length() {
+    return crimson::os::seastore::get_encoded_record_length(record, block_size);
+  }
+  size_t get_wouldbe_encoded_record_length(LogicalCachedExtentRef& extent) {
+    auto raw_mdlength = get_encoded_record_raw_mdlength(record, block_size);
+    auto wouldbe_mdlength = p2roundup(
+      raw_mdlength + ceph::encoded_sizeof_bounded<extent_info_t>(),
+      block_size);
+    return wouldbe_mdlength + extent_buf_len + extent->get_bptr().length();
+  }
+  ceph::bufferlist encode(segment_id_t segment, segment_nonce_t nonce) {
+    assert(extents.size() == record.extents.size());
+    auto rsize = get_encoded_record_length();
+    segment_off_t extent_offset = base + rsize.mdlength;
+    for (auto& extent : extents) {
+      extent->set_ool_paddr(
+        {segment, extent_offset});
+      extent_offset += extent->get_bptr().length();
+    }
+    return encode_record(rsize, std::move(record), block_size, nonce);
+  }
+  void add_extent(LogicalCachedExtentRef& extent) {
+    extents.emplace_back(new OolExtent(extent));
+    ceph::bufferlist bl;
+    bl.append(extent->get_bptr());
+    record.extents.emplace_back(extent_t{
+      extent->get_type(),
+      extent->get_laddr(),
+      std::move(bl)});
+    extent_buf_len += extent->get_bptr().length();
+  }
+  std::vector<OolExtentRef>& get_extents() {
+    return extents;
+  }
+  void set_base(segment_off_t b) {
+    base = b;
+  }
+  segment_off_t get_base() {
+    return base;
+  }
+  void clear() {
+    record.extents.clear();
+    extents.clear();
+    assert(!record.deltas.size());
+    extent_buf_len = 0;
+    base = MAX_SEG_OFF;
+  }
+  uint64_t get_num_extents() {
+    return extents.size();
+  }
+private:
+  std::vector<OolExtentRef> extents;
```

Comment:
Why not store the extents directly?

Reply:
I thought we'd better set the extents' poffset to the absolute address after their content is persisted to disk, so the absolute address has to be kept somewhere. Since this is an internal procedure of the EPM, it seems not appropriate to let the CachedExtent hold that absolute address temporarily, so I added this "OolExtent" wrapper just to hold that ool absolute addr.

Reply:
Mmmm, these extents are private to the transaction until commit (new extents, aren't in the general cache yet).  OolExtent doesn't seem necessary and you're storing them by pointer necessitating an extra allocation.  I think you can just store the extents directly and update the address as soon as you have it or wait until after commit and derive the addresses from the record base address and the location of the extents in the record vector.

Reply:
Here is a related comment: https://github.com/ceph/ceph/pull/41803#discussion_r678931346

Previously `ool_offset` was implemented as an additional field in `CachedExtent` which I think is not necessary as @xxhdx1985126 also pointed out "this is an internal procedure of the EPM".

And I agree with @athanatos that `OolExtent` can be more lightweighted and doesn't need to be dynamically allocated.

Reply:
Oh, sorry, there was another reason for this OolExtent wrapper. I preserved the old paddr check in `LBAMananger::update_mapping()` from the original `LBAManager::rewrite_extent()`, which requires that EPM keep an extent's old paddr somewhere. So I created that OolExtent to keep the old extent untouched until the lba mapping update happens

Reply:
> Mmmm, these extents are private to the transaction until commit (new extents, aren't in the general cache yet). OolExtent doesn't seem necessary and you're storing them by pointer necessitating an extra allocation. I think you can just store the extents directly and update the address as soon as you have it or wait until after commit and derive the addresses from the record base address and the location of the extents in the record vector.

It seems that, to derive the address from the location of the extents in the record vector, we still have to keep that location somewhere. So it looks to me that to drop OolExtent, we can only update the poffset of the extent directly right before the segment write. And then again, if we eventually decide to hold that old paddr check in `LBAManager::update_mapping()`, we still have to keep the old paddr of the extents somewhere, which means we may still need something like the OolExtent. Am I right? I tend to preserve that old paddr check as it seems important for validating the correctness of future development.

> doesn't need to be dynamically allocated.

I can do this for a more lightweighted OolExtent.

Reply:
@athanatos @cyx1231st I've drop the new allocation of OolExtent, please take a look again, thanks:-)